README.md
# Solace Integration: RDP to Azure Functions

[![tests](https://github.com/solace-iot-team/solace-int-rdp-az-funcs/actions/workflows/tests.yml/badge.svg)](https://github.com/solace-iot-team/solace-int-rdp-az-funcs/actions/workflows/tests.yml)

Integrations of Solace with Azure Functions using the Solace Rest Delivery Point (RDP).

[Issues](https://github.com/solace-iot-team/solace-int-rdp-az-funcs/issues) |
[Release Notes](./ReleaseNotes.md) |

## solace-rdp-2-blob

Azure Function integrating Solace RDP to Azure Storage (Blob).

Sequence:
- message ==> Solace Broker
- Solace Broker Rest Delivery Point ==> Azure Function (solace-rdp-2-blob)
- Azure Function ==> Azure Storage


### Azure Storage

- Account kind: StorageV2
- Hierarchical namespace: enabled

### Azure Function
[See also test directory for Build, Deploy, Test.](./test)

**Configuration:**

[See ./solace-rdp-2-blob/template.app.settings.json](./solace-rdp-2-blob/template.app.settings.json)

* **Rdp2BlobStorageContainerName**
  - container name
* **Rdp2BlobStoragePathPrefix**
  - fixed path prefix for each file
* **Rdp2BlobStorageConnectionString**
  - connection string for the data lake storage account


**URI Query Parameters:**

* **path**="{level-1/level-2/level-3/...}" - string, can be empty
  - multi-level path for the file (message) in the Blob Storage
  - appended to the _Rdp2BlobStoragePathPrefix_ configuration setting

* **pathCompose**=["withTime"], optional
  - _withTime_
    - generates a time-based storage path: YYYY/MM/DD/MM
    - appended to _Rdp2BlobStoragePathPrefix/path_

**Ouptut:**

|http code|description|
|---|---|
|200   | ok, file written to storage  |
|400   | bad request. return for config errors & query parameter errors  |
|4xx   | pass-through of Azure blob storage REST Api  |

Console log:
- ERROR details in case of any errors


## Dev & Test

[See test directory.](./test)

---


ReleaseNotes.md
# Release Notes

## Version: 0.2.0
Unit & integration testing.

### function: solace-rdp-2-blob
* changed Azure configuration setting names to:
  ````json
  {
    "Rdp2BlobStorageContainerName": "{dataLakeStorageContainerName}",
    "Rdp2BlobStoragePathPrefix": "{dataLakeFixedPathPrefix}",
    "Rdp2BlobStorageConnectionString": "{dataLakeStorageConnectionString}"
  }
  ````

## Version: 0.1.0
Initial Release.

---


solace-rdp-2-blob/index.ts

// Copyright (c) 2020, Solace Corporation, Ricardo Gomez-Ulmke (ricardo.gomez-ulmke@solace.com).
// All rights reserved.
// Licensed under the MIT License.

/*
TODO: TESTS
    - container does not exist --> created?

    Mocha & Chai, sinonjs for mocking and stubbing properties.
    https://adrianhall.github.io/web/2018/07/04/run-typescript-mocha-tests-in-vscode/

    https://github.com/Testy/TestyTs

*/


import { AzureFunction, Context, HttpRequest } from "@azure/functions"
import { BlobServiceClient } from "@azure/storage-blob" 
import { generateUuid, RestError } from "@azure/core-http"
import { FunctionArgs, ArgItem, ArgSpec } from "../solace-rdp-lib/FunctionArgs"
import { ConfigError } from "../solace-rdp-lib/Errors"

function composeBlobName(appSettings: FunctionArgs, queryParams: FunctionArgs): string {

    if (queryParams.getValue(queryParamPathCompose) === pathComposeWithTime) {
        let timeStamp = new Date();
        let month: string = String(timeStamp.getUTCMonth() + 1).padStart(2, '0');
        let day: string = String(timeStamp.getUTCDate()).padStart(2, '0');
        let hours: string = String(timeStamp.getUTCHours() + 1).padStart(2, '0');
        let minutes: string = String(timeStamp.getUTCMinutes()).padStart(2, '0');
        let seconds: string = String(timeStamp.getUTCSeconds()).padStart(2, '0');
        return `${appSettings.getValue(appSettingStoragePathPrefix)}/${queryParams.getValue(queryParamPath)}/${timeStamp.getUTCFullYear()}/${month}/${day}/${hours}/${minutes}/${seconds}_${generateUuid()}`
    }
    return `${appSettings.getValue(appSettingStoragePathPrefix)}/${queryParams.getValue(queryParamPath)}/${generateUuid()}`
}

const appSettingStorageConnectionString: string = "Rdp2BlobStorageConnectionString";
const appSettingStoragePathPrefix: string = "Rdp2BlobStoragePathPrefix";
const appSettingStorageContainerName: string = "Rdp2BlobStorageContainerName";
const appSettingsSpec: ArgSpec = [
    new ArgItem(appSettingStorageConnectionString),
    new ArgItem(appSettingStoragePathPrefix),
    new ArgItem(appSettingStorageContainerName)
]
const queryParamPath: string = "path";
const queryParamPathCompose: string = "pathCompose";
const pathComposeWithTime: string = "withTime";
const queryParamsSpec: ArgSpec = [
    new ArgItem(queryParamPath, true, null, null, null),
    new ArgItem(queryParamPathCompose, false, "none", ["none", pathComposeWithTime])
]

/**
 * https://www.npmjs.com/package/@azure/storage-blob
 * https://docs.microsoft.com/en-gb/javascript/api/@azure/storage-blob/?view=azure-node-latest
 * https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-error-codes
 * @param context 
 * @param req 
 */
export const solaceRDP2Blob: AzureFunction = async function (context: Context, req: HttpRequest): Promise<void> {

    context.log.info(`[INFO] - STARTING: ${context.executionContext.functionName} ...`);
    context.log.info(`[INFO] - req=${JSON.stringify(req, null, 2)}`);

    context.log.info(`[INFO] - appSettingsSpec=${JSON.stringify(appSettingsSpec)}`)
    context.log.info(`[INFO] - queryParamsSpec=${JSON.stringify(queryParamsSpec)}`)

    try {
        const appSettings = new FunctionArgs('app-settings', process.env, appSettingsSpec);
        const queryParams = new FunctionArgs('query-params', req.query, queryParamsSpec);
        context.log.info(`[INFO] - settings=${appSettings.toString()}`);
        context.log.info(`[INFO] - queryParams=${queryParams.toString()}`);

        let blobServiceClient: BlobServiceClient = null;
        try {
            blobServiceClient = BlobServiceClient.fromConnectionString(appSettings.getValue(appSettingStorageConnectionString));
        } catch (err) {
            throw new ConfigError(`'appSettings.${appSettingStorageConnectionString}': ${err.message}`);
        }
        const containerName = appSettings.getValue(appSettingStorageContainerName);
        const containerClient = blobServiceClient.getContainerClient(containerName);
        const blobName = composeBlobName(appSettings, queryParams);
        const blockBlobClient = containerClient.getBlockBlobClient(blobName);
        const content = req.rawBody;
        if ( content === undefined || content.length === 0) {
            context.log.warn(`[WARN] - message has no content, discarding.`);
            context.res = { status: 200 }; 
            return;
        }
        try {
            const uploadBlobResponse = await blockBlobClient.upload(content, content.length);
        } catch(err) {
            if(err instanceof RestError) {
                let details: any = err.details;
                if(err.statusCode === 404 && details.errorCode === 'ContainerNotFound') {
                    // create the container first
                    const createContainerResponse = await containerClient.create();
                    context.log.info(`[INFO] - container created. name:'${containerName}'`);
                    // reject message ==> will receive message again    
                    context.res = { status: 400 };  
                    return;
                } else {
                    context.log.error(`[ERROR] - uploading block blob failed: statusCode: ${err.statusCode}, details: ${JSON.stringify(err.details)}`);
                    if (err.statusCode === 400 && details.errorCode === 'InvalidResourceName') {                    
                        throw new ConfigError(`${details.message}\n(possible cause: invalid container name: '${containerName}')`);
                    } else {
                        throw err; // uknown RestError
                    }
                }
            } else if(err instanceof TypeError){
                context.log.error(`[ERROR] - uploading block blob failed: details: ${err.message}`);
                throw new ConfigError(`${err.message}\n(possible cause: invalid blob connection string: '${appSettings.getValue(appSettingStorageConnectionString)}')`);
            } else throw err;
        } 

        context.log.info(`[INFO] - upload block blob success: '${blobName}'`);

        context.res = { status: 200 }; 

    } catch(err) {
        let status = 400; // default: Bad Request
        if( err instanceof RestError) {
            status = err.statusCode;
        } else if( err instanceof ConfigError) { 
            status = 400; // any better code?
        }
        context.log.error(`[ERROR] - ${err}`);        
        context.res = { status: status };
    }

};

// The End.